[ADAM-710] Add saveAs methods for feature formats GTF, BED, IntervalList, and NarrowPeak

heuermh opened the pull request:
Fixes #710.

Pushing this early for review.

Still todo is writing IntervalList header and ```-single``` file support, if that might be desirable.  Currently this uses Spark ```saveAsTextFile``` which writes to as many ```part-00...``` text files as there are partitions.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1149/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 72f238f7ce10833095118399dda8b2c36a85c2ae automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/998/merge^{commit} # timeout=10 > git branch -a --contains 385dbc23bcb4b849c4303829347615c78cacdb0e # timeout=10 > git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 385dbc23bcb4b849c4303829347615c78cacdb0e (origin/pr/998/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 385dbc23bcb4b849c4303829347615c78cacdb0eFirst time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1152/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 3630a2670b18045807f503c5fd2ebe2cff282e1b automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/998/merge^{commit} # timeout=10 > git branch -a --contains 7d86cfe3d4f8646f59969530f775bc58976348b5 # timeout=10 > git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 7d86cfe3d4f8646f59969530f775bc58976348b5 (origin/pr/998/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 7d86cfe3d4f8646f59969530f775bc58976348b5First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1153/<h2>Build result: FAILURE</span></h2>[...truncated 24 lines...]Triggering ADAM-prb ? 2.6.0,2.10,1.3.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.6.0,centosTriggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.5.2,centosADAM-prb ? 2.6.0,2.11,1.6.0,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.5.2,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.6.0,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.6.0,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.10,1.3.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.6.0,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.3.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.3.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.3.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.10,1.5.2,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1157/
Test PASSed.

heuermh commented:
@tdanford Some specific review questions for you, if you have the time:

 - Is there a bug here? https://github.com/heuermh/adam/blob/feature-writers/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala#L188
 - Can we add ```frame``` as a optional field to the schema?  Currently I'm pushing it in and out of attributes https://github.com/heuermh/adam/blob/feature-writers/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala#L162
 - Why the UUID for ```featureId```?  https://github.com/heuermh/adam/blob/feature-writers/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala#L226  It makes round-tripping not possible
 - Should we add Dbxref support to all the formats?  Currently only IntervalListParser parses them out https://github.com/heuermh/adam/blob/feature-writers/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala#L226

heuermh commented:
@tdanford bump for feedback, see last comment

fnothaft commented:
@heuermh what `needs work` on this PR? Do you think we can get this in to 0.20.0 (end of month)?

heuermh commented:
The four items in the comment above that need review and single-file SAM/BAM-style header support for IntervalList similar to that described in #1009.

fnothaft commented:
OK. I'll get going on the review.

> single-file SAM/BAM-style header support for IntervalList similar to that described in #1009.

What does the `IntervalList` header look like? I'll work on #1009 and we can then slot this on top.

heuermh commented:
> What does the `IntervalList` header look like?

"The file format is relatively simple and reflects the SAM alignment format to a degree. A SAM style header must be present in the file which lists the sequence records against which the intervals are described."

https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/util/IntervalList.html

Thus we might be able to reuse the SAM header stuff directly; the current IntervalList [parsing code](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala#L133) does not.

fnothaft commented:
> Thus we might be able to reuse the SAM header stuff directly; the current IntervalList parsing code does not. You had mentioned refactoring the SAM/BAM header code although maybe not as part of #1009; I can pick this up based on what you do there.

I would prefer using the SAM header code to rolling our own.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1237/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 35af436d7eb02df20af5d6c4b5f190f0bafb7a16 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 95b97512b3abd38a07c7f6fc97460bea315cb338 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 95b97512b3abd38a07c7f6fc97460bea315cb338 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 95b97512b3abd38a07c7f6fc97460bea315cb338First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1238/
Test PASSed.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -118,7 +122,11 @@ class GTFParser extends FeatureParser {
 
-    f.setAttributes(attrs)
+    if (frame != ".") {
+      f.setAttributes(collection.mutable.Map(attrs.toSeq: _*) += ("frame" -> frame))
```
`collection.mutable.Map(attrs.toSeq: _*) += ("frame" -> frame)` to `attrs.map(v -> ("frame", v)).toMap)`?

fnothaft replied:
Also, could we get some docs here? What's this conditional for? It looks like we're special casing some formatting thing. Why's that? Etc.

heuermh replied:
Yeah I don't like it, I will propose adding frame as an attribute to Feature in an upcoming pull request.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -96,6 +96,10 @@ class GTFParser extends FeatureParser {
       .setFeatureType(feature)
       .setSource(source)
 
+    if (score != ".") {
```
Ditto here RE: adding docs for special case code.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -177,6 +185,7 @@ class IntervalListParser extends Serializable {
             Feature.newBuilder()
               .setContig(Contig.newBuilder().setContigName(fields(0)).build())
               .setStart(fields(1).toLong - 1)
+              // todo: IntervalList ranges are closed, is this correct?
```
This is correct. Drop `todo` but keep comment as documentation? E.g., `// NOTE: IntervalList ranges are closed`

heuermh replied:
If IntervalList intervals are 1-based and closed, shouldn't the next line be
```scala
.setEnd(fields(2).toLong)
```
E.g. `[7-9]` in IntervalList should be `[6-9)` in our 0-based, closed open system.

fnothaft replied:
Agreed, my `correct` was RE: the comment.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -196,8 +205,16 @@ class IntervalListParser extends Serializable {
 class BEDParser extends FeatureParser {
 
+  def isHeader(line: String): Boolean = {
+    return line.startsWith("#") || line.startsWith("browser") || line.startsWith("track")
```
Nit: Drop `return`

heuermh replied:
What, you want my Scala not to look like Java?  :)

fnothaft replied:
;)

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -196,8 +205,16 @@ class IntervalListParser extends Serializable {
   override def parse(line: String): Seq[Feature] = {
 
+    if (isHeader(line)) {
+      return Seq()
```
Ditto here. I'd rather refactor the function so that we eliminate all `return` statements.

fnothaft replied:
Ping.

heuermh replied:
here I find it better to short circuit return than to put the whole method in an if block

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -230,22 +247,22 @@ class BEDParser extends FeatureParser {
     }
     val attributes = new ArrayBuffer[(String, String)]()
-    if (fields.length > 6) {
+    if (fields.length > 6 && fields(6) != ".") {
```
It seems like there's a lot of repeated code here. Could we refactor? E.g.:

```
def attrExists(attrIdx: Int, attr: String) {
  if (fields.length > attrIdx && fields(attrIdx) != ".") {
    attributes += ("attr" -> fields(attrIdx)
  }
}

attrExists(6, "thickStart")
attrExists(7, "thickEnd")
...
```

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala:
```diff
@@ -140,4 +141,96 @@ class FeatureRDDFunctions(featureRDD: RDD[Feature]) extends Serializable with Lo
+        Joiner.on('\t').join(chrom, start, end, name, score, strand)
+      } else {
+        // write BED12 format
+        val thickStart = feature.getAttributes.getOrElse("thickStart", ".")
```
I would factor this out --> `feature.getAttributes.getOrElse(attr, ".")` to:

```
def getAttr(attr: String): String = {
  feature.getAttributes.getOrElse(attr, ".")
}

val thickStart = getAttr("thickStart")
...
```

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala:
```diff
@@ -140,4 +141,96 @@ class FeatureRDDFunctions(featureRDD: RDD[Feature]) extends Serializable with Lo
+        case Strand.Reverse     => "-"
+        case Strand.Independent => "."
+      }
+      val signalValue = feature.getAttributes.getOrElse("signalValue", "0")
```
Might factor these out too.

fnothaft commented on adam-core/src/test/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctionsSuite.scala:
```diff
@@ -0,0 +1,256 @@
+import org.bdgenomics.formats.avro._
+import org.scalactic.{ Equivalence, TypeCheckedTripleEquals }
+
+object FeatureRDDFunctionsSuiteHelper {
```
The functions in this object seem like they'd be useful to surface outside of the `test` scope.

fnothaft commented on adam-core/src/test/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctionsSuite.scala:
```diff
@@ -0,0 +1,256 @@
+}
+
+class FeatureRDDFunctionsSuiteHelper extends Serializable {
+  // empty
```
What's this here for?

heuermh replied:
Adds the Serializable marker, otherwise tests will error

fnothaft replied:
You can just add that to the singleton object, no? E.g., `object FeatureRDDFunctionsSuiteHelper extends Serializable`

fnothaft commented on adam-core/src/test/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctionsSuite.scala:
```diff
@@ -0,0 +1,256 @@
+  }
+
+  /*
+  sparkTest("round trip IntervalList format") {
```
Eliminate dead test code, or change to an ignored test --> `ignore("round trip..."){`

fnothaft commented:
LGTM! I just made a pass through with a variety of nits.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1242/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 8c212da69ee941b25ede9896c3250de81c7d6de9 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 4f03eb49e355afd1765b413d1f1442ea02dac6e0 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 4f03eb49e355afd1765b413d1f1442ea02dac6e0 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 4f03eb49e355afd1765b413d1f1442ea02dac6e0First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

fnothaft commented:
@heuermh when do you think we can wrap this by?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1252/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit dc491aecaf2b6eddef6319bf5f391c67a27efee4 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains f1ad9fbc852ed49d66a938228e210a2aeb1c8b00 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision f1ad9fbc852ed49d66a938228e210a2aeb1c8b00 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f f1ad9fbc852ed49d66a938228e210a2aeb1c8b00First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

heuermh commented:
Updated with GFF3 support and addressed some comments above.

Still left to do:
 * Continue refactoring `FeatureParser`to reuse functionality where appropriate and improve consistency between formats
 * Refactor `toGenes` to use new proposed feature fields (`geneId`, `transcriptId`, and `exonId`) instead of attributes
 * `-single` file support, if that might be desirable
 * Writing header for IntervalList and GFF3 formats

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1255/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 2f260e4f154c314081242fbcd12e8dd16c613ebe automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 7783adc282593f152f6a034b7c5371f54a3bf45b # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 7783adc282593f152f6a034b7c5371f54a3bf45b (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 7783adc282593f152f6a034b7c5371f54a3bf45bFirst time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1256/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit fd68ccc30375a044c9005d2f76a8e4c232257ee1 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 511ddd9484d32b5198010244f33a1fe975170df6 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 511ddd9484d32b5198010244f33a1fe975170df6 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 511ddd9484d32b5198010244f33a1fe975170df6First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1257/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit fd1e8e23363e3d981c96193cf2ce4ad6e2f73928 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains d0eff44fdac1d0d2459d8100ae5b3e806517679c # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision d0eff44fdac1d0d2459d8100ae5b3e806517679c (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f d0eff44fdac1d0d2459d8100ae5b3e806517679cFirst time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1260/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit fa0f17fc9d29ccefe9f2c452f4d0b73d2ae7a6a3 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains d07a7e8f1e210681c3d25fb1eade7a21350822a6 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision d07a7e8f1e210681c3d25fb1eade7a21350822a6 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f d07a7e8f1e210681c3d25fb1eade7a21350822a6First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1267/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit ae89357cd5344a349c57228b5b818fe9a0039832.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse ae89357cd5344a349c57228b5b818fe9a0039832^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains ae89357cd5344a349c57228b5b818fe9a0039832 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/head^{commit} # timeout=10Checking out Revision ae89357cd5344a349c57228b5b818fe9a0039832 (origin/pr/998/head) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f ae89357cd5344a349c57228b5b818fe9a0039832First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1268/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit aa5b8676b82adcbf4397b6943451d763881a57e3 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains a0e77fb619019140b03bfdf6aa3e9f60dc1dee92 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision a0e77fb619019140b03bfdf6aa3e9f60dc1dee92 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f a0e77fb619019140b03bfdf6aa3e9f60dc1dee92First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1269/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 65d3d22cca6f3872f3e4cd75204ca4a6d9f49a96 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains f0d387b9f2719548c7d80103add98614b21dfbb8 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision f0d387b9f2719548c7d80103add98614b21dfbb8 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f f0d387b9f2719548c7d80103add98614b21dfbb8 > /home/jenkins/git2/bin/git rev-list a0e77fb619019140b03bfdf6aa3e9f60dc1dee92 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1270/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 8e2d7864aadde1f199b8007fb48843e618316eba automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 2579282e083482e1d781ec622231a24f601e1aa1 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 2579282e083482e1d781ec622231a24f601e1aa1 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 2579282e083482e1d781ec622231a24f601e1aa1 > /home/jenkins/git2/bin/git rev-list f0d387b9f2719548c7d80103add98614b21dfbb8 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

heuermh commented:
Ready for another round of review.  I'd be ok pushing this and adding `-single` and header support later.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -44,86 +46,120 @@ class FeatureFile(parser: FeatureParser) extends Serializable {
+}
+
+object GFF3Parser {
+  def parseAttributes(attributes: String): Seq[(String, String)] =
```
Add docs? This is a weird line. I don't love the pattern match on the array, but that's just me. I would probably do something like:

```
attributes.split(";")
  .map(s => {
    val eqIdx = s.indexOf("=")
    (s.take(eqIdx), s.drop(eqIdx + 1))
  }).toSeq
```

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -44,86 +46,120 @@ class FeatureFile(parser: FeatureParser) extends Serializable {
-      attrs.get("transcript_id").flatMap(t => attrs.get("exon_number").map(e => t + "_" + e))
+    val fields = line.split("\t")
+    // skip empty or invalid lines
+    if (fields.length < 9) {
```
Should log invalid lines.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -44,86 +46,120 @@ class FeatureFile(parser: FeatureParser) extends Serializable {
+
+    // set score if specified
+    if (score != ".") {
+      f.setScore(source.toDouble)
```
s/source/score/g?

heuermh replied:
the unit tests should've caught this typo!

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -194,58 +230,65 @@ class IntervalListParser extends Serializable {
     }
     if (fields.length > 4) {
-      fb.setValue(fields(4) match {
+      fb.setScore(fields(4) match {
```
I don't love this pattern match. I would just:

```
if (fields(4) != ".") {
  fb.setScore(fields(4).toDouble)
}
```

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala:
```diff
@@ -194,58 +230,65 @@ class IntervalListParser extends Serializable {
     }
     val attributes = new ArrayBuffer[(String, String)]()
-    if (fields.length > 6) {
+    if (fields.length > 6 && fields(6) != ".") {
```
Factor out `fields.length > n && fields(n) != "."` into:

```
def validIdx(n: Int): Boolean = {
  fields.length > n && fields(n) != "."
}
```

Or something like that.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala:
```diff
@@ -133,11 +132,167 @@ class FeatureRDDFunctions(featureRDD: RDD[Feature]) extends Serializable with Lo
+        case None    => transcriptId.flatMap(t => exonNumber.map(e => t + "_" + e))
+      }
+
+      val (_featureId, _parentId) =
```
We don't use `_` prefixes elsewhere in this codebase, so I'd prefer not to introduce them here. I would suffix `featureId: Option[String]` to `featureIdOpt`, which is fairly common across Scala projects.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala:
```diff
@@ -133,11 +132,167 @@ class FeatureRDDFunctions(featureRDD: RDD[Feature]) extends Serializable with Lo
+  }
+
+  def saveAsIntervalList(fileName: String) = {
+    def escape(entry: (Any, Any)): String = {
```
Where does `escape` get used? Perhaps I'm just missing things but I don't see it get called anywhere.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala:
```diff
@@ -133,11 +132,167 @@ class FeatureRDDFunctions(featureRDD: RDD[Feature]) extends Serializable with Lo
+        f.getEnd,
+        f.getFeatureType,
+        // try exon_id first
+        Option(f.getExonId).getOrElse("~"),
```
Is there any significance to `~`? I would document.

heuermh replied:
do nulls sort first or last in scala?  I'm used to e.g. Ordering.nullsFirst in Guava.

fnothaft replied:
Not sure. I was figuring that this was to handle the null case. If so, can you add a line comment to indicate so?

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDDFunctions.scala:
```diff
@@ -133,11 +132,167 @@ class FeatureRDDFunctions(featureRDD: RDD[Feature]) extends Serializable with Lo
+        Option(f.getExonId).getOrElse("~"),
+        // then try transcript_id and exon_number
+        Option(f.getTranscriptId).getOrElse("~"),
+        f.getAttributes.getOrElse("exon_number", Int.MaxValue.toString).toInt
```
Prefer `f.getAttributes.get("exon_number").map(_.toInt).getOrElse(Int.MaxValue)`

heuermh replied:
nulls don't work in scala Ordering, and this line doesn't compile or handle nulls

fnothaft replied:
Ah, sorry I mentally had the return type of `f.getAttributes` as a Scala Map. `f.getAttributes` returns a `java.util.String`, right? If you convert this return value into a Scala Map (e.g., via [`scala.collection.JavaConversions`](http://www.scala-lang.org/api/2.10.4/index.html#scala.collection.JavaConversions$) or etc), `.get("exon_number")` returns an `Option[String]`, and then the `.getOrElse` handles the `null` case, no?

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+package org.bdgenomics.adam.rdd.features
+
+import scala.collection.JavaConversions._
+
```
Nit: whitespace.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+import scala.collection.mutable.HashMap
+import scala.collection.mutable.MutableList
+
+import org.bdgenomics.formats.avro.{ Dbxref, Feature, OntologyTerm, Strand }
```
Nit: import sort order --> `org` before `scala`.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+   * @param strand strand to convert
+   * @return the specified strand converted to its string value
+   */
+  def toString(strand: Strand): String = {
```
I would avoid `toString` as a function name for Singleton objects.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+
+import scala.collection.JavaConversions._
+
+import scala.collection.mutable.ArrayBuffer
```
Prefer `import scala.collection.mutable.{...}` to imports from same package across multiple lines.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+   * @param dbxref database cross reference to convert
+   * @return the specified database cross reference converted to its string value
+   */
+  def toString(dbxref: Dbxref): String = {
```
Ditto RE naming of `toString`.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+   * @param ontologyTerm ontology term to convert
+   * @return the specified ontology term converted to its string value
+   */
+  def toString(ontologyTerm: OntologyTerm): String = {
```
Ditto

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    // set remaining attributes if non-empty;
+    // any duplicate keys are lost at this point, last one in wins
```
Would it make sense to log duplicate keys?

heuermh replied:
+1, will add logging

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+      attrs += Tuple2("Is_circular", b.toString)
+    }
+
+    // ... and why Tuple2 is required here
```
OOC, does it compile OK if you don't use an anon argument? E.g., does `Option(feature.getFeatureId).foreach(id => attrs += ("ID", id))` compile?

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+    Option(feature.getGeneId).foreach(attrs += Tuple2("gene_id", _))
+    Option(feature.getTranscriptId).foreach(attrs += Tuple2("transcript_id", _))
+    Option(feature.getExonId).foreach(attrs += Tuple2("exon_id", _))
+    for (alias <- feature.getAliases) attrs += Tuple2("Alias", alias)
```
I generally prefer `feature.getAliases.foreach` to `for (...)`, but that's just me.

heuermh replied:
I don't think that worked before, maybe for the same anon argument issue you mention above.  Will give it a try.

fnothaft replied:
You'd need to convert the `java.util.List` into a Scala collection for it to work.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala:
```diff
@@ -0,0 +1,238 @@
+   * @return a name for the specified feature considering its attributes
+   */
+  def nameOf(feature: Feature): String = {
+    Option(feature.getName).foreach(return _)
```
I strongly dislike explicit returns in Scala code. I would rather:

```
def nameOf(feature: Feature): String = {
  Option(feature.getName).getOrElse(
    Option(feature.getFeatureId).getOrElse(feature.getFeatureType match {
      case "exon" => Option(feature.getExonId)
      case "transcript" => Option(feature.getTranscriptId)
      case "gene" => Option(feature.getGeneId)
      case s: String => Some(s)
      case _ => None
    }.getOrElse("feature")))
}
```

heuermh replied:
this and various permutations doesn't work for me; compile errors or `value getOrElse is not a member of java.io.Serializable`

fnothaft commented:
Just made a pass. Overall, LGTM! Dropped various style nits and the like inline.

> I'd be ok pushing this and adding -single and header support later.

I +1.

Also, this adds all the code needed for handling the changes in https://github.com/bigdatagenomics/bdg-formats/pull/82, correct?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1271/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 0948bf2abc23581ea5242998e75929a1463d39fb automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 3bdd3ca832b784d31332b73b82a804f2acb2c718 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 3bdd3ca832b784d31332b73b82a804f2acb2c718 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 3bdd3ca832b784d31332b73b82a804f2acb2c718 > /home/jenkins/git2/bin/git rev-list 2579282e083482e1d781ec622231a24f601e1aa1 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

heuermh commented:
Addressed most review comments and rebased.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1272/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit fd433347b89fe4fe580d55011348818d984f7c29 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 446f4707d04574442db241eab65bb484835fcbd0 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 446f4707d04574442db241eab65bb484835fcbd0 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 446f4707d04574442db241eab65bb484835fcbd0 > /home/jenkins/git2/bin/git rev-list 3bdd3ca832b784d31332b73b82a804f2acb2c718 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

fnothaft commented:
Jenkins, test this please.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1273/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit fd433347b89fe4fe580d55011348818d984f7c29 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 446f4707d04574442db241eab65bb484835fcbd0 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 446f4707d04574442db241eab65bb484835fcbd0 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 446f4707d04574442db241eab65bb484835fcbd0 > /home/jenkins/git2/bin/git rev-list 446f4707d04574442db241eab65bb484835fcbd0 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

heuermh commented:
The unit test failure is due to indeterminate sorting. Is there a way to easily sort by deep compare (nested arrays and maps)?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1274/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit a2a688c91a3c2617d3d9076bede72264eb1c835d automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 0660916dbadb4f13232ff50ed053042689f8b380 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 0660916dbadb4f13232ff50ed053042689f8b380 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 0660916dbadb4f13232ff50ed053042689f8b380 > /home/jenkins/git2/bin/git rev-list 446f4707d04574442db241eab65bb484835fcbd0 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

heuermh commented:
Maybe that's not the problem.  I don't see this error locally.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1277/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit 1c1fe235c51838cd20d5cf2fe13455ab053a407a automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 95ea3c23796024c70a81ce78474b6601a9690ce6 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 95ea3c23796024c70a81ce78474b6601a9690ce6 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 95ea3c23796024c70a81ce78474b6601a9690ce6 > /home/jenkins/git2/bin/git rev-list 0660916dbadb4f13232ff50ed053042689f8b380 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1278/<h2>Build result: FAILURE</span></h2>[...truncated 24 lines...]Triggering ADAM-prb ? 2.6.0,2.10,1.6.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.6.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.6.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.4.1,centosADAM-prb ? 2.6.0,2.11,1.6.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.10,1.3.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.10,1.6.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.6.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.5.2,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.5.2,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.6.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.4.1,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

heuermh commented:
Jenkins, test this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1279/
Test PASSed.

fnothaft commented:
[bdg-formats 0.9.0 has pushed](http://search.maven.org/#artifactdetails%7Corg.bdgenomics.bdg-formats%7Cbdg-formats%7C0.9.0%7Cjar).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1287/<h2>Build result: FAILURE</span></h2>GitHub pull request #998 of commit d78910d428e73a2143f47864d08b161cf39bbdb6 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > /home/jenkins/git2/bin/git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > /home/jenkins/git2/bin/git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > /home/jenkins/git2/bin/git --version # timeout=10 > /home/jenkins/git2/bin/git -c core.askpass=true fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* # timeout=15 > /home/jenkins/git2/bin/git rev-parse origin/pr/998/merge^{commit} # timeout=10 > /home/jenkins/git2/bin/git branch -a --contains 8ca998ac3574053ba0d7e9425dcab7e524006228 # timeout=10 > /home/jenkins/git2/bin/git rev-parse remotes/origin/pr/998/merge^{commit} # timeout=10Checking out Revision 8ca998ac3574053ba0d7e9425dcab7e524006228 (origin/pr/998/merge) > /home/jenkins/git2/bin/git config core.sparsecheckout # timeout=10 > /home/jenkins/git2/bin/git checkout -f 8ca998ac3574053ba0d7e9425dcab7e524006228 > /home/jenkins/git2/bin/git rev-list 317561754912b3f81306fd1c564265f027e5e574 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.5.2,centosTriggering ADAM-prb ? 2.6.0,2.11,1.5.2,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

fnothaft commented:
Jenkins, retest this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/1288/
Test PASSed.

fnothaft commented:
Thanks @heuermh! I've squashed this down to a single commit and merged as https://github.com/bigdatagenomics/adam/commit/8abb47ad2447bbc0376f9b5052e2e029bdf6b604.

heuermh commented:
Thanks!
